### PR TITLE
Replacing depreciated add() for #1099

### PR DESCRIPTION
**Pull request by iqbalabd**
I've tried replacing calls for `add()` with `register()` when using Markdown.

### CODE REVIEW

**Comment by benjaoming** [comment]
Thanks so so much for the contribution! It seems that Travis CI has decided to do something disruptive. So there is currently no CI. I will have to investigate. It's not the first time that they have done this, but we did get approved as an Open Source project earlier this year.

**Comment by benjaoming** [comment]
Okay, the integration is perhaps back now. Passing by quickly from a bad connection so couldn't get to push a commit to this PR, but would you mind pushing something... fingers crossed that the Travis build will then start!

**Comment by benjaoming** [comment]
Yeah this is clearly more than I could handle in a short time. We were not on Travis CI, but did get offered the Open Source plan earlier. Had made a migration to Circle CI. But for no apparent reason it stopped working after #1110

**Comment by benjaoming** [comment]
Okay I have no idea about the sorcery here - can you try pulling in the latest `main` branch? Thanks so much :pray:

**Comment by iqbalabd** [comment]
I've tried pulling the latest main branch and can see the 4 new updates.

**Comment by benjaoming** [comment]
There must be some sort of security mechanism I've overlooked. #1126 is building just fine.

Could you try pushing to your `issue-1099` branch just to check?

**Comment by iqbalabd** [comment]
Hmm. I tried merging from upstream (again) and pushing it again, and it seems that the CI is working right now. I'm not really sure what happened the first time.

File: src/wiki/core/markdown/mdx/codehilite.py [hunk 1]
```diff
@@ -107,7 +107,15 @@ def extendMarkdown(self, md):
                 "'codehilite' from WIKI_MARKDOWN_KWARGS"
             )
             del md.treeprocessors["hilite"]
-        md.treeprocessors.add("hilite", hiliter, "<inline")
+
+        i = md.treeprocessors.get_index_for_name("inline")
+        after = md.treeprocessors._priority[i].priority
+        if i > 0:
+            before = md.treeprocessors._priority[i-1].priority
```
Comment (benjaoming) [inline src/wiki/core/markdown/mdx/codehilite.py:114]:
```suggestion
            before = md.treeprocessors._priority[i - 1].priority
```

File: src/wiki/core/markdown/mdx/codehilite.py [hunk 2]
```diff
@@ -117,7 +125,15 @@ def extendMarkdown(self, md):
             del md.preprocessors["fenced_code_block"]
         hiliter = WikiFencedBlockPreprocessor(md)
         hiliter.config = self.getConfigs()
-        md.preprocessors.add("fenced_code_block", hiliter, ">normalize_whitespace")
+
+        i = md.preprocessors.get_index_for_name("normalize_whitespace")
+        before = md.preprocessors._priority[i].priority
+        if i < len(md.preprocessors) - 1:
+            after = md.preprocessors._priority[i+1].priority
```
Comment (benjaoming) [inline src/wiki/core/markdown/mdx/codehilite.py:132]:
```suggestion
            after = md.preprocessors._priority[i + 1].priority
```

File: src/wiki/core/markdown/mdx/responsivetable.py [hunk 3]
```diff
@@ -7,7 +7,10 @@ class ResponsiveTableExtension(markdown.Extension):
     """Wraps all tables with Bootstrap's table-responsive class"""
 
     def extendMarkdown(self, md):
-        md.treeprocessors.add("responsivetable", ResponsiveTableTree(md), "_end")
+
+        md.treeprocessors._sort()
+        priority = md.treeprocessors._priority[-1].priority - 5
```
Comment (benjaoming) [inline src/wiki/core/markdown/mdx/responsivetable.py:12]: This looks like the best solution, given the options. It's a pity that wanting to put an extension at the end isn't a very clearly defined argument.

File: src/wiki/plugins/attachments/markdown_extensions.py [hunk 4]
```diff
@@ -19,9 +19,15 @@ class AttachmentExtension(markdown.Extension):
...
     def extendMarkdown(self, md):
         """ Insert AbbrPreprocessor before ReferencePreprocessor. """
-        md.preprocessors.add(
-            "dw-attachments", AttachmentPreprocessor(md), ">html_block"
-        )
+
+        i = md.preprocessors.get_index_for_name("html_block")
+        before = md.preprocessors._priority[i].priority
+        if i < len(md.preprocessors) - 1:
+            after = md.preprocessors._priority[i+1].priority
```
Comment (benjaoming) [inline src/wiki/plugins/attachments/markdown_extensions.py:26]:
```suggestion
            after = md.preprocessors._priority[i + 1].priority
```

File: src/wiki/plugins/images/markdown_extensions.py [hunk 5]
```diff
@@ -30,8 +30,24 @@ class ImageExtension(markdown.Extension):
...
+            after = md.inlinePatterns._priority[i+1].priority
+        else:
+            after = before - 10
+        priority = before - ((before - after) / 2)
+        md.inlinePatterns.register(ImagePattern(IMAGE_RE, md), "dw-images", priority)
+
+        i = md.postprocessors.get_index_for_name("raw_html")
+        before = md.postprocessors._priority[i].priority
+        if i < len(md.postprocessors) - 1:
+            after = md.postprocessors._priority[i+1].priority
```
Comment (benjaoming) [inline src/wiki/plugins/images/markdown_extensions.py:46]:
```suggestion
            after = md.postprocessors._priority[i + 1].priority
```

File: src/wiki/plugins/macros/mdx/macro.py [hunk 6]
```diff
@@ -20,7 +20,15 @@ class MacroExtension(markdown.Extension):
     """ Macro plugin markdown extension for django-wiki. """
 
     def extendMarkdown(self, md):
-        md.inlinePatterns.add("dw-macros", MacroPattern(MACRO_RE, md), ">link")
+
+        i = md.inlinePatterns.get_index_for_name("link")
+        before = md.inlinePatterns._priority[i].priority
+        if i < len(md.inlinePatterns) - 1:
+            after = md.inlinePatterns._priority[i+1].priority
```
Comment (benjaoming) [inline src/wiki/plugins/macros/mdx/macro.py:27]:
```suggestion
            after = md.inlinePatterns._priority[i + 1].priority
```

File: src/wiki/plugins/macros/mdx/wikilinks.py [hunk 7]
```diff
@@ -41,7 +41,15 @@ def extendMarkdown(self, md):
         WIKILINK_RE = r"\[\[([\w0-9_ -]+)\]\]"
         wikilinkPattern = WikiLinks(WIKILINK_RE, self.getConfigs())
         wikilinkPattern.md = md
-        md.inlinePatterns.add("wikilink", wikilinkPattern, "<not_strong")
+
+        i = md.inlinePatterns.get_index_for_name("not_strong")
+        after = md.inlinePatterns._priority[i].priority
+        if i > 0:
+            before = md.inlinePatterns._priority[i-1].priority
```
Comment (benjaoming) [inline src/wiki/plugins/macros/mdx/wikilinks.py:48]:
```suggestion
            before = md.inlinePatterns._priority[i - 1].priority
```

File: src/wiki/plugins/redlinks/mdx/redlinks.py [hunk 8]
```diff
@@ -97,7 +97,15 @@ def extendMarkdown(self, md):
         md.registerExtension(self)
         self.md = md
         ext = self.TreeProcessorClass(md, self.getConfigs())
-        md.treeprocessors.add("redlinks", ext, ">inline")
+
+        i = md.treeprocessors.get_index_for_name("inline")
+        before = md.treeprocessors._priority[i].priority
+        if i < len(md.treeprocessors) - 1:
+            after = md.treeprocessors._priority[i+1].priority
```
Comment (benjaoming) [inline src/wiki/plugins/redlinks/mdx/redlinks.py:104]:
```suggestion
            after = md.treeprocessors._priority[i + 1].priority
```

**Review by benjaoming — APPROVED** [review]
Thanks for having a look at this. It clearly gives a lot of repetitive patterns, but it should really be fixed in Python-Markdown and not here, so I think the consequence that this logic now lives here makes sense, given that Python-Markdown now handles less wrt. extension priority :+1:

File: src/wiki/plugins/images/markdown_extensions.py [hunk 9]
```diff
@@ -30,8 +30,24 @@ class ImageExtension(markdown.Extension):
     """ Images plugin markdown extension for django-wiki. """
 
     def extendMarkdown(self, md):
-        md.inlinePatterns.add("dw-images", ImagePattern(IMAGE_RE, md), ">link")
-        md.postprocessors.add("dw-images-cleanup", ImagePostprocessor(md), ">raw_html")
+
+        i = md.inlinePatterns.get_index_for_name("link")
+        before = md.inlinePatterns._priority[i].priority
+        if i < len(md.inlinePatterns) - 1:
+            after = md.inlinePatterns._priority[i+1].priority
```
Comment (benjaoming) [inline src/wiki/plugins/images/markdown_extensions.py:37]:
```suggestion
            after = md.inlinePatterns._priority[i + 1].priority
```

File: src/wiki/plugins/attachments/markdown_extensions.py [hunk 10]
```diff
@@ -19,9 +19,15 @@ class AttachmentExtension(markdown.Extension):
...
-        )
+
+        i = md.preprocessors.get_index_for_name("html_block")
+        before = md.preprocessors._priority[i].priority
+        if i < len(md.preprocessors) - 1:
+            after = md.preprocessors._priority[i + 1].priority
+        else:
+            after = before - 10
+        priority = before - ((before - after) / 2)
+        md.preprocessors.register(AttachmentPreprocessor(md), "dw-attachments", priority)
```
Comment (benjaoming) [inline src/wiki/plugins/attachments/markdown_extensions.py:30]:
```suggestion
        md.preprocessors.register(
            AttachmentPreprocessor(md), "dw-attachments", priority
        )
```

File: src/wiki/plugins/images/markdown_extensions.py [hunk 11]
```diff
@@ -30,8 +30,24 @@ class ImageExtension(markdown.Extension):
...
+        md.inlinePatterns.register(ImagePattern(IMAGE_RE, md), "dw-images", priority)
+
+        i = md.postprocessors.get_index_for_name("raw_html")
+        before = md.postprocessors._priority[i].priority
+        if i < len(md.postprocessors) - 1:
+            after = md.postprocessors._priority[i + 1].priority
+        else:
+            after = before - 10
+        priority = before - ((before - after) / 2)
+        md.postprocessors.register(ImagePostprocessor(md), "dw-images-cleanup", priority)
```
Comment (benjaoming) [inline src/wiki/plugins/images/markdown_extensions.py:50]:
```suggestion
        md.postprocessors.register(
            ImagePostprocessor(md), "dw-images-cleanup", priority
        )
```

**Comment by iqbalabd** [comment]
Thanks for approving @benjaoming . Perhaps refactoring the repeating code into separate functions might make it look better, but being my first commit to the project, I'm not familiar enough with it, so I did the most obvious.

What should I do now? Do I leave this pull request as it is and it will be merged later?

**Comment by benjaoming** [comment]
Hi @iqbalabd - First of all, thanks for the contribution and for fixing the issue! The contribution is valid and stands as an overall improvement on itself. So I will merge it now. But your suggestion to perhaps build a utility function seems like a good idea.

I can imagine that it could live here? https://github.com/django-wiki/django-wiki/blob/main/src/wiki/core/markdown/__init__.py

If you are able to fit another PR for this, it's most welcome :pray:

**Comment by iqbalabd** [comment]
Thanks for accepting the PR @benjaoming . I will look into refactoring the changes next, based on your suggestion.

**Pull request merged**